src/mame/machine/k573fpga.h

```diff
@@ -20,12 +20,10 @@ class k573fpga_device : public device_t
 	void set_ram(u16 *v) { ram = v; }
 	u16 get_decrypted();
 
-	void set_crypto_key1(u16 v) { crypto_key1 = v; }
+	void set_crypto_key1(u16 v) { orig_crypto_key1 = crypto_key1 = v; }
 	void set_crypto_key2(u16 v) { crypto_key2 = v; }
 	void set_crypto_key3(u8 v) { crypto_key3 = v; }
-	u16 get_crypto_key1() const { return crypto_key1; }
-	u16 get_crypto_key2() const { return crypto_key2; }
-	u8 get_crypto_key3() const { return crypto_key3; }
+	u16 get_crypto_key1() const { return orig_crypto_key1; }
```

Comment:
I don't remember the fpga storing the original key.  I'll have to check.

Reply:
Check Dance Dance Revolution Solo Bass Mix. That's the only game that I know of that was strict about the keys. If I remember correctly it was something along the lines of the key is checked before it starts a song, and it must be equal to the key that it was set to be (or something like that). It was the only game I could find that actually cared about the contents of the key registers though.

A few months ago I wrote a hacky homebrew tool to decrypt data using the FPGA directly using Drummania 5th Mix as a base so I could easily modify that to verify this on real hardware.

However, you're probably right that it's not technically the *FPGA* that's storing the original key, so this particular piece of code should probably be moved to k573dio.

Reply:
For the time being, I moved the copy of the original key1 to k573dio. 

I remembered I had some pics of memory contents during decryption where I dumped the 0x1f640000-0x1f640100 range and double checked but I actually don't see the keys anywhere in that data. I'm not very confident about that bit anymore. It might actually be the FPGA returning that data, and it might be something specific to DDR Solo Bass Mix because DDR Solo Bass Mix is the only special game on the platform with its unique FPGA program.